src/slurm_plugin/clustermgtd.py

```diff
@@ -0,0 +1,624 @@
+        try:
+            # Set nodes into down state so jobs can be requeued immediately
+            set_nodes_down(node_list, reason="Static node maintenance: unhealthy node is being replaced")
+        except Exception as e:
+            logging.error(
+                "Error setting unhealthy static nodes into down state, continuing with instance replacement: %s", e,
+            )
+        instances_to_terminate = []
+        for node in unhealthy_static_nodes:
+            backing_instance = private_ip_to_instance_map.get(node.nodeaddr)
+            if backing_instance:
+                instances_to_terminate.append(backing_instance.id)
+        log.info(
+            "Terminating instances %s that are backing unhealthy static nodes: %s",
+            instances_to_terminate,
+            unhealthy_static_nodes,
+        )
+        self.instance_manager.delete_instances(
+            instances_to_terminate, terminate_batch_size=self.sync_config.terminate_max_batch_size
+        )
+        log.info("Launching new instances for unhealthy static nodes: %s", unhealthy_static_nodes)
+        self.instance_manager.add_instances_for_nodes(
+            node_list, self.sync_config.launch_max_batch_size, self.sync_config.update_node_address
+        )
+        # Add node to list of nodes being replaced
+        self.static_nodes_in_replacement |= set(node_list)
+        log.info(
+            "After node maintenance, following nodes are currently in replacement: %s", self.static_nodes_in_replacement
+        )
+
+    def _maintain_nodes(self, cluster_instances, slurm_nodes):
+        """
+        Call functions to maintain unhealthy nodes.
+
+        This function needs to handle the case that 2 slurm nodes have the same IP/nodeaddr.
+        Hence the a list of nodes is passed in and slurm node dict with IP/nodeaddr as key should be avoided.
+        """
+        try:
+            # Update self.static_nodes_in_replacement by removing any up nodes from the set
+            self._update_static_nodes_in_replacement(slurm_nodes)
+            log.info("Following nodes are currently in replacement: %s", self.static_nodes_in_replacement)
+            private_ip_to_instance_map = {instance.private_ip: instance for instance in cluster_instances}
+            unhealthy_dynamic_nodes, unhealthy_static_nodes = self._find_unhealthy_slurm_nodes(
+                slurm_nodes, private_ip_to_instance_map
+            )
+            log.info(
+                "Found the following unhealthy static nodes: %s\nFound the following unhealthy dynamic nodes: %s",
+                unhealthy_static_nodes,
+                unhealthy_dynamic_nodes,
+            )
+            if unhealthy_dynamic_nodes:
+                ClusterManager._handle_unhealthy_dynamic_nodes(unhealthy_dynamic_nodes)
+            if unhealthy_static_nodes:
+                self._handle_unhealthy_static_nodes(unhealthy_static_nodes, private_ip_to_instance_map)
+        except Exception as e:
```

Comment:
do we want that a failure in _handle_unhealthy_dynamic_nodes also affects _handle_unhealthy_static_nodes?